### PR TITLE
Potential Optimizations

### PR DESCRIPTION
**Pull request by rmorshea**
Using the same simple test as #463 I was able to improve performance by ~30% in contrast to #463's ~40% boost. However this PR doesn't introduce caching, and instead registers descriptors via `subclass_init` to `HasDescriptors._descriptors`.  Ultimately this means that neither `HasTraits.traits`, nor `setup_instance` traverse all of `dir(cls)` to find traits and descriptors.

### CODE REVIEW

File: traitlets/traitlets.py [hunk 1]
```diff
@@ -791,6 +792,7 @@ def setup_class(cls, classdict):
         BaseDescriptor in the class dict of the newly created ``cls`` before
         calling their :attr:`class_init` method.
         """
+        cls._descriptors = []
```
Comment (minrk) [inline traitlets/traitlets.py]: Does this new state need to go in the `__getstate__`?

**Comment by jf---** [comment]
A 30% performance increase is quite the accomplishment... given the obvious upside, why not merge things?

File: traitlets/traitlets.py (hunk 1 above)
Comment (rmorshea) [reply traitlets/traitlets.py]: @minrk if I'm not mistaken you don't have to recreate state for classes  when pickling so `_descriptors` and `_trait_default_generators` shouldn't need to be in `__getstate__`. When the module containing these classes is reloaded then these fields should be repopulated automatically.

**Comment by rmorshea** [comment]
@jf--- responded to @minrk and awaiting his approval.

**Comment by minrk** [comment]
@rmorshea sorry this repo is so neglected.

Can you add a test that covers pickle/unpickle and dynamically-defined instance traits with `add_traits`? That would make me more sanguine about moving a bunch of state to the class-only level.

**Comment by Carreau** [comment]
Closing as state for a few years, apologies for not getting this in. I want to avoid having multiple page on stale PRs on this repository.

Labelling accordingly feel free to reopen.

**Pull request closed without merging**